File: src/recurrence.ts
import { invoke } from '@tauri-apps/api/core';
import type { RecurrenceKind, Transaction, MerchantOverride, Category } from './types';

export interface RecurrenceInfo {
  kind: RecurrenceKind;
  count: number;
  meanIntervalDays: number;
  meanAmount: number;
  estMonthlyCost: number;
  lastDate: string | null;
  source: 'auto' | 'override';
}

export async function buildRecurrenceMap(isDemo: boolean = false): Promise<Record<string, RecurrenceInfo>> {
  try {
    return await invoke<Record<string, RecurrenceInfo>>('build_recurrence_map', { isDemo });
  } catch (e) {
    console.error("Failed to build recurrence map natively:", e);
    return {};
  }
}

export function isRecurring(kind: RecurrenceKind): boolean {
  return kind !== 'none';
}

export function recurrenceLabel(kind: RecurrenceKind): string {
  switch (kind) {
    case 'monthly':
      return 'Monthly';
    case 'biweekly':
      return 'Biweekly';
    case 'weekly':
      return 'Weekly';
    case 'annual':
      return 'Annual';
    case 'none':
      return 'One-time';
  }
}

export async function refreshRecurrenceAll(): Promise<{ updated: number }> {
  try {
    return await invoke<{ updated: number }>('refresh_recurrence_all');
  } catch (e) {
    console.error("Failed to refresh recurrence natively:", e);
    return { updated: 0 };
  }
}

const NONE: RecurrenceInfo = {
  kind: 'none',
  count: 0,
  meanIntervalDays: 0,
  meanAmount: 0,
  estMonthlyCost: 0,
  lastDate: null,
  source: 'auto',
};

function daysBetween(a: string, b: string): number {
  return Math.round(
    (new Date(b).getTime() - new Date(a).getTime()) / 86_400_000
  );
}

export function detectRecurrence(txns: Transaction[]): RecurrenceInfo {
  const spend = txns
    .filter((t) => t.amount < 0)
    .slice()
    .sort((a, b) => a.date.localeCompare(b.date));

  if (spend.length < 3) {
    return { ...NONE, count: spend.length, lastDate: spend.at(-1)?.date ?? null };
  }

  const intervals: number[] = [];
  for (let i = 1; i < spend.length; i++) {
    intervals.push(daysBetween(spend[i - 1].date, spend[i].date));
  }
  const mean = intervals.reduce((s, x) => s + x, 0) / intervals.length;
  const variance =
    intervals.reduce((s, x) => s + (x - mean) ** 2, 0) / intervals.length;
  const stddev = Math.sqrt(variance);

  let kind: RecurrenceKind = 'none';

  const meanAmt = spend.reduce((s, t) => s + Math.abs(t.amount), 0) / spend.length;
  const amtVariance = spend.reduce((s, t) => s + (Math.abs(t.amount) - meanAmt) ** 2, 0) / spend.length;
  const amtStddev = Math.sqrt(amtVariance);
  const isStableAmount = amtStddev <= (meanAmt * 0.3) + 5; 

  if (isStableAmount) {
    if (mean >= 25 && mean <= 35 && stddev <= 7) kind = 'monthly';
    else if (mean >= 12 && mean <= 17 && stddev <= 5) kind = 'biweekly';
    else if (mean >= 5 && mean <= 9 && stddev <= 3) kind = 'weekly';
    else if (mean >= 330 && mean <= 400 && stddev <= 40) kind = 'annual';
  }

  const recentSpend = spend.slice(-3);
  const meanAmount =
    recentSpend.reduce((s, t) => s + Math.abs(t.amount), 0) /
    recentSpend.length;

  let estMonthlyCost = 0;
  switch (kind) {
    case 'monthly':
      estMonthlyCost = meanAmount;
      break;
    case 'biweekly':
      estMonthlyCost = meanAmount * (30 / 14);
      break;
    case 'weekly':
      estMonthlyCost = meanAmount * (30 / 7);
      break;
    case 'annual':
      estMonthlyCost = meanAmount / 12;
      break;
  }

  return {
    kind,
    count: spend.length,
    meanIntervalDays: mean,
    meanAmount,
    estMonthlyCost,
    lastDate: spend[spend.length - 1].date,
    source: 'auto',
  };
}

export function resolveRecurrenceForTransaction(
  txn: Omit<Transaction, 'id' | 'recurrence'> & { recurrenceOverride?: 'recurring' | 'onetime' | null },
  categoryMap: Map<string, Category>,
  merchantOverrideMap: Map<string, any>,
  autoRecurringMerchantKeys: Set<string>
): 'recurring' | 'onetime' {
  if (txn.recurrenceOverride === 'recurring') return 'recurring';
  if (txn.recurrenceOverride === 'onetime') return 'onetime';
  const mkey = txn.merchantKey;
  if (mkey) {
    const override = merchantOverrideMap.get(mkey);
    if (override) {
      return override.recurrence === 'none' || override.recurrence === 'onetime' ? 'onetime' : 'recurring';
    }
  }
  if (mkey && autoRecurringMerchantKeys.has(mkey)) {
    return 'recurring';
  }
  const cat = categoryMap.get(txn.category);
  if (cat?.defaultRecurrence) {
    return cat.defaultRecurrence;
  }
  return 'onetime';
}

File: src-tauri/src/tools/recurrence.rs
use rusqlite::Result;
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use chrono::NaiveDate;
use tauri::State;
use crate::db::{DbState, Transaction, MerchantOverride, Category};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RecurrenceInfo {
    pub kind: String,
    pub count: usize,
    #[serde(rename = "meanIntervalDays")]
    pub mean_interval_days: f64,
    #[serde(rename = "meanAmount")]
    pub mean_amount: f64,
    #[serde(rename = "estMonthlyCost")]
    pub est_monthly_cost: f64,
    #[serde(rename = "lastDate")]
    pub last_date: Option<String>,
    pub source: String,
}

impl Default for RecurrenceInfo {
    fn default() -> Self {
        Self {
            kind: "none".to_string(),
            count: 0,
            mean_interval_days: 0.0,
            mean_amount: 0.0,
            est_monthly_cost: 0.0,
            last_date: None,
            source: "auto".to_string(),
        }
    }
}

fn days_between(a: &str, b: &str) -> i64 {
    let date_a = NaiveDate::parse_from_str(a, "%Y-%m-%d").unwrap_or_default();
    let date_b = NaiveDate::parse_from_str(b, "%Y-%m-%d").unwrap_or_default();
    (date_b - date_a).num_days()
}

pub fn detect_recurrence(txns: &[&Transaction]) -> RecurrenceInfo {
    let mut spend: Vec<&Transaction> = txns.iter().filter(|t| t.amount < 0.0).copied().collect();
    spend.sort_by(|a, b| a.date.cmp(&b.date));

    if spend.len() < 3 {
        let mut info = RecurrenceInfo::default();
        info.count = spend.len();
        info.last_date = spend.last().map(|t| t.date.clone());
        return info;
    }

    let mut intervals = Vec::new();
    for i in 1..spend.len() {
        intervals.push(days_between(&spend[i-1].date, &spend[i].date) as f64);
    }
    
    let mean = intervals.iter().sum::<f64>() / intervals.len() as f64;
    let variance = intervals.iter().map(|x| (x - mean).powi(2)).sum::<f64>() / intervals.len() as f64;
    let stddev = variance.sqrt();

    let mean_amt = spend.iter().map(|t| t.amount.abs()).sum::<f64>() / spend.len() as f64;
    let amt_variance = spend.iter().map(|t| (t.amount.abs() - mean_amt).powi(2)).sum::<f64>() / spend.len() as f64;
    let amt_stddev = amt_variance.sqrt();
    let is_stable_amount = amt_stddev <= (mean_amt * 0.3) + 5.0;

    let mut kind = "none".to_string();
    if is_stable_amount {
        if mean >= 25.0 && mean <= 35.0 && stddev <= 7.0 { kind = "monthly".to_string(); }
        else if mean >= 12.0 && mean <= 17.0 && stddev <= 5.0 { kind = "biweekly".to_string(); }
        else if mean >= 5.0 && mean <= 9.0 && stddev <= 3.0 { kind = "weekly".to_string(); }
        else if mean >= 330.0 && mean <= 400.0 && stddev <= 40.0 { kind = "annual".to_string(); }
    }

    let recent_spend = if spend.len() >= 3 { &spend[spend.len()-3..] } else { &spend[..] };
    let mean_amount = recent_spend.iter().map(|t| t.amount.abs()).sum::<f64>() / recent_spend.len() as f64;

    let est_monthly_cost = match kind.as_str() {
        "monthly" => mean_amount,
        "biweekly" => mean_amount * (30.0 / 14.0),
        "weekly" => mean_amount * (30.0 / 7.0),
        "annual" => mean_amount / 12.0,
        _ => 0.0,
    };

    RecurrenceInfo {
        kind,
        count: spend.len(),
        mean_interval_days: mean,
        mean_amount,
        est_monthly_cost,
        last_date: Some(spend.last().unwrap().date.clone()),
        source: "auto".to_string(),
    }
}

pub fn apply_override(auto: RecurrenceInfo, overide: Option<&MerchantOverride>) -> RecurrenceInfo {
    if let Some(o) = overide {
        let kind = o.recurrence.clone();
        let mut est_monthly_cost = auto.est_monthly_cost;
        if kind != auto.kind && auto.mean_amount > 0.0 {
            est_monthly_cost = match kind.as_str() {
                "monthly" => auto.mean_amount,
                "biweekly" => auto.mean_amount * (30.0 / 14.0),
                "weekly" => auto.mean_amount * (30.0 / 7.0),
                "annual" => auto.mean_amount / 12.0,
                "none" => 0.0,
                _ => auto.est_monthly_cost,
            };
        }
        RecurrenceInfo {
            kind,
            est_monthly_cost,
            source: "override".to_string(),
            ..auto
        }
    } else {
        auto
    }
}

pub fn build_recurrence_map_internal(txns: &[Transaction], overrides: &[MerchantOverride]) -> HashMap<String, RecurrenceInfo> {
    let mut by_merchant: HashMap<String, Vec<&Transaction>> = HashMap::new();
    for t in txns {
        if !t.merchant_key.is_empty() {
            by_merchant.entry(t.merchant_key.clone()).or_default().push(t);
        }
    }

    let override_map: HashMap<String, &MerchantOverride> = overrides.iter()
        .map(|o| (o.merchant_key.clone(), o))
        .collect();

    let mut result = HashMap::new();
    for (merchant_key, m_txns) in by_merchant {
        let recurring_txns: Vec<&Transaction> = m_txns.iter().filter(|t| t.recurrence == "recurring").copied().collect();
        if recurring_txns.is_empty() { continue; }

        let spend_txns: Vec<&Transaction> = m_txns.iter().filter(|t| t.amount < 0.0).copied().collect();
        let auto = detect_recurrence(&spend_txns);
        let mut final_info = apply_override(auto, override_map.get(&merchant_key).copied());

        if final_info.kind == "none" {
            final_info.kind = "monthly".to_string();
            let mean_amount = recurring_txns.iter().map(|t| t.amount.abs()).sum::<f64>() / recurring_txns.len() as f64;
            final_info.mean_amount = mean_amount;
            final_info.est_monthly_cost = mean_amount;
        }

        final_info.count = recurring_txns.len();
        final_info.last_date = recurring_txns.last().map(|t| t.date.clone());

        result.insert(merchant_key, final_info);
    }

    result
}

#[tauri::command]
pub fn build_recurrence_map(state: State<DbState>, is_demo: bool) -> Result<HashMap<String, RecurrenceInfo>, String> {
    if !is_demo {
        let cache = state.recurrence_cache.lock().unwrap();
        if let Some(cached_map) = cache.as_ref() {
            return Ok(cached_map.clone());
        }
    }

    let txns = crate::db::get_transactions(state.clone(), "1970-01-01".into(), "2099-12-31".into(), None, None, None, None).map_err(|e| e.to_string())?;
    let overrides = crate::db::get_merchant_overrides(state.clone()).map_err(|e| e.to_string())?;

    let filtered_txns: Vec<Transaction> = txns.into_iter()
        .filter(|t| if is_demo { t.source == "demo" } else { t.source != "demo" })
        .collect();

    let map = build_recurrence_map_internal(&filtered_txns, &overrides);

    if !is_demo {
        let mut cache = state.recurrence_cache.lock().unwrap();
        *cache = Some(map.clone());
    }

    Ok(map)
}

pub fn resolve_recurrence_for_transaction(
    txn: &Transaction,
    category_map: &HashMap<String, &Category>,
    merchant_override_map: &HashMap<String, &MerchantOverride>,
    auto_recurring_merchant_keys: &HashSet<String>,
) -> String {
    if let Some(ro) = &txn.recurrence_override {
        if ro == "recurring" || ro == "onetime" {
            return ro.clone();
        }
    }

    let mkey = &txn.merchant_key;
    if !mkey.is_empty() {
        if let Some(over) = merchant_override_map.get(mkey) {
            if over.recurrence == "none" || over.recurrence == "onetime" {
                return "onetime".to_string();
            } else {
                return "recurring".to_string();
            }
        }
    }

    if !mkey.is_empty() && auto_recurring_merchant_keys.contains(mkey) {
        return "recurring".to_string();
    }

    if let Some(cat) = category_map.get(&txn.category) {
        if let Some(def_rec) = &cat.default_recurrence {
            return def_rec.clone();
        }
    }

    "onetime".to_string()
}

#[derive(Serialize)]
pub struct RefreshResult {
    pub updated: usize,
}

#[tauri::command]
pub fn refresh_recurrence_all(state: State<DbState>) -> Result<RefreshResult, String> {
    let categories = crate::db::get_categories(state.clone()).map_err(|e| e.to_string())?;
    let overrides = crate::db::get_merchant_overrides(state.clone()).map_err(|e| e.to_string())?;
    let mut txns = crate::db::get_transactions(state.clone(), "1970-01-01".into(), "2099-12-31".into(), None, None, None, None).map_err(|e| e.to_string())?;

    let category_map: HashMap<String, &Category> = categories.iter().map(|c| (c.name.clone(), c)).collect();
    let merchant_override_map: HashMap<String, &MerchantOverride> = overrides.iter().map(|o| (o.merchant_key.clone(), o)).collect();

    let mut by_merchant: HashMap<String, Vec<&Transaction>> = HashMap::new();
    for t in &txns {
        if !t.merchant_key.is_empty() {
            by_merchant.entry(t.merchant_key.clone()).or_default().push(t);
        }
    }

    let mut auto_recurring_merchant_keys = HashSet::new();
    for (mkey, m_txns) in by_merchant {
        let auto = detect_recurrence(&m_txns);
        if auto.kind != "none" {
            auto_recurring_merchant_keys.insert(mkey);
        }
    }

    let mut updated = 0;
    let mut txs_to_update = Vec::new();

    for t in &mut txns {
        let resolved = resolve_recurrence_for_transaction(
            t,
            &category_map,
            &merchant_override_map,
            &auto_recurring_merchant_keys,
        );
        if t.recurrence != resolved {
            t.recurrence = resolved;
            txs_to_update.push(t.clone());
            updated += 1;
        }
    }

    if !txs_to_update.is_empty() {
        crate::db_mut::bulk_update_transactions(state.clone(), txs_to_update).map_err(|e| e.to_string())?;
    }

    Ok(RefreshResult { updated })
}
